src/main/java/com/github/cbl/algorithm_analyzer/trees/AvlTree/AVLTree.java

```diff
@@ -0,0 +1,432 @@
+
+        Node<T> right();
+
+        boolean empty();
+
+        Node<T> insert(T t);
+
+        Node<T> remove(T t);
+
+        Optional<? extends Node<T>> find(T t);
+
+        Optional<Integer> depth(T t);
+
+        default int height() {
+            return 1 + Math.max(left().height(), right().height());
+        }
+
+        Node<T> rebalance();
+
+        Node<T> rotateRight();
+
+        Node<T> rotateLeft();
+
+        default int balance() {
+            return right().height() - left().height();
+        }
+
+        default T min() {
+            T m = left().min();
+            return null != m ? m : value();
+        }
+
+        default T successor() {
+            return right().min();
+        }
+    }
+
+    /****************************
+     *     Node implementation  *
+     ****************************/
+
+    private class TreeNode implements Node<T> {
+        private final T value;
+        private final Node<T> left;
+        private final Node<T> right;
+
+        public TreeNode(T value, Node<T> left, Node<T> right) {
+            assert (value != null);
+            this.value = value;
+            this.left = left;
+            this.right = right;
+        }
+
+        @Override
+        public T value() {
+            return value;
+        }
+
+        @Override
+        public Node<T> right() {
+            return right;
+        }
+
+        @Override
+        public Node<T> left() {
+            return left;
+        }
+
+        @Override
+        public boolean empty() {
+            return false;
+        }
+
+        @Override
+        public TreeNode insert(T value) {
+            int cmp = AVLTree.this.comparator.compare(value, this.value);
+            if (cmp < 0) {
+                return new TreeNode(this.value, left().insert(value), right()).rebalance();
+            } else if (cmp > 0) {
+                return new TreeNode(this.value, left(), right().insert(value)).rebalance();
+            } else {
+                return this;
+            }
+        }
+
+        @Override
+        public Node<T> remove(T value) {
+            int cmp = AVLTree.this.comparator.compare(value, this.value);
+            if (cmp < 0) {
+                return new TreeNode(this.value, left().remove(value), right()).rebalance();
+            } else if (cmp > 0) {
+                return new TreeNode(this.value, left(), right().remove(value)).rebalance();
+            } else {
+                if (!left().empty() && !right().empty()) {
+                    T succ = this.successor();
+                    return new TreeNode(succ, left(), right().remove(succ)).rebalance();
+                } else if (right().empty()) {
+                    return left();
+                } else {
+                    return right();
+                }
+            }
+        }
+
+        @Override
+        public Optional<? extends Node<T>> find(T value) {
+            int cmp = AVLTree.this.comparator.compare(value, this.value);
```

Comment:
Just out of interest, would this work the sameß
```suggestion
            int cmp = this.comparator.compare(value, this.value);
```

Reply:
No it wouldn't, I just noticed that `this` would reference to the instance of the `TreeNode`.

Reply:
Yep, [nested classes](https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html) are are a special kind of sorcery 🧙‍♂️